IDR/production/1.compute_statistics.py

```diff
@@ -0,0 +1,102 @@
+    for key, value in elements_and_counts_dict.items():
+        # Check if value is of dict type
+        if isinstance(value, dict):
+            # If value is dict then iterate over all its values
+            for pair in nested_dict_pairs_iterator(value):
+                yield (key, *pair)
+        else:
+            # If value is not dict type then yield the value
+            yield (key, value)
+
+
+if __name__ == "__main__":
+    # Define study metadata directory
+    studies_metadata_dir = pathlib.Path("IDR/data/metadata")
+
+    # Collect metadata file paths
+    metadata_files = list(walk(studies_metadata_dir))
+    metadata_files = metadata_files[0:5]
+
+    # Make directories
+    stats_dir = pathlib.Path("IDR/data/statistics")
+    pathlib.Path.mkdir(stats_dir, exist_ok=True)
+
+    # Calculate statistics for each image attribute for individual studies
+    individual_study_stats = list()
+    unique_elements_and_counts = dict()
+
+    # Iterate through each study/screen/well.json metadata file
+    for metadata_path in metadata_files:
+        study_name = ((str(metadata_path).split("/")[-1]).split(".")[0]).split("_")[0]
+        screen_id = ((str(metadata_path).split("/")[-1]).split(".")[0]).split("_")[-1]
+        _, attribute_elements = collect_study_stats(
+            metadata_path,
+            individual_study_stats,
+            na_cols=["screen_id", "study_name", "plate_name", "plate_id", "well_id"],
+            study_name=study_name,
+            screen_id=screen_id,
+        )
+
+        unique_elements_and_counts[study_name] = attribute_elements
+
+    # Convert unique_elements_and_counts dict to Pandas.DataFrame
+    unique_elements_and_counts_list = list(
+        nested_dict_pairs_iterator(unique_elements_and_counts)
+    )
+
+    # Generate output dataframes
+    unique_elements_and_counts_df = pd.DataFrame(
+        data=unique_elements_and_counts_list,
+        columns=["Study", "Screen", "Attribute", "Element", "n"],
+    )
+    stat_results_df = pd.DataFrame(
+        data=individual_study_stats,
+        columns=["Study_Name", "Attribute", "S", "H", "NME", "J", "E", "GC"],
+    )
+
+    # Save elements and counts as parquet file
+    elements_and_counts_output_file = pathlib.Path(
+        stats_dir, "unique_elements_and_counts.parquet"
+    )
+    unique_elements_and_counts_df.to_parquet(elements_and_counts_output_file)
+
+    # Save individual stats as parquet file
+    indv_studies_output_file = pathlib.Path(
+        stats_dir, "individual_studies_diversity.parquet"
+    )
+    stat_results_df.to_parquet(indv_studies_output_file)
+
+    # Collect databank stats
+    databank_stats = collect_databank_stats(
+        metadata_directory=studies_metadata_dir, na_cols=[]
+    )
+
+    # Save databank stats as parquet file
+    databank_output_file = pathlib.Path(stats_dir, f"databank_diversity.parquet")
+    databank_stats.to_parquet(databank_output_file)
```

Comment:
I'm a bit confused how you're collecting the full databank_stats when you're also collecting individual study stats. Don't you need results from multiple studies to collect the full databank stats?

Reply:
@gwaybio I collect the individual stats by iterating through each metadata file. For the databank stats, I concatenate all of the study metadata file contents and compute from one big metadata dataframe.